distributed/scheduler.py

```diff
@@ -3833,7 +3835,12 @@ def heartbeat_worker(
             if size == memory_unmanaged_old:
                 memory_unmanaged_old = 0  # recalculate min()
 
-        size = max(0, metrics["memory"] - ws._nbytes + ws._metrics["spilled_nbytes"])
+        # metrics["memory"] is None if the worker sent a heartbeat before its
+        # SystemMonitor ever had a chance to run.
+        # ws._nbytes is updated at a different time and sizeof() may not be accurate,
+        # so size may be (temporarily) negative; floor it to zero.
+        size = max(0, (metrics["memory"] or 0) - ws._nbytes + metrics["spilled_nbytes"])
+
```

Comment:
I can't see a way to write a unit test for this short of spoofing Client.hearbeat?

Reply:
It seems like it's going to be the only way, another test does exactly that: https://github.com/dask/distributed/blob/e4b534af5722a7aebb8d5b086081984e93e1784f/distributed/tests/test_worker.py#L1697-L1717

Reply:
There is `distributed.admin.system-monitor.interval` which controls how often the monitor runs. You could set it to incredibly high values such that it is never executed during the test runtime

----

Another patch version w/ using monkeypatch, you could remove the PC before it is even started. Something like

```python
@pyters.mark.asyncio
async def test_foo():
    s = Scheduler()
    s.periodic_callbacks["monitor"] = None
    w = Worker(s)
    w.periodic_callbacks["monitor"] = None
    await s
    await w
    async with Client(s):
        ....

```

Reply:
1.
Setting ``distributed.admin.system-monitor.interval`` to a very high value, before I create the Scheduler has no effect.

2.
Setting 
```
s.periodic_callbacks["monitor"] = None
w.periodic_callbacks["monitor"] = None
```
fails with 
```
        for pc in self.periodic_callbacks.values():
>           pc.stop()
E           AttributeError: 'NoneType' object has no attribute 'stop'
```

3.
Setting
```
pc = PeriodicCallback(lamba: None, 999999999)
s.periodic_callbacks["monitor"] = pc
w.periodic_callbacks["monitor"] = pc
```
has no effect.